### PR TITLE
Restore Form URL encoding support.

### PR DESCRIPTION
**Pull request by JakeWharton**
@swankjesse @mattprecious

### CODE REVIEW

File: retrofit/src/main/java/retrofit/FormEncodingBuilder.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+      throw new AssertionError(e);
+    }
+    content.writeUtf8(name);
+    content.writeByte('=');
+    content.writeUtf8(value);
+    return this;
+  }
+
+  RequestBody build() {
+    if (content.size() == 0) {
```
Comment (swankjesse) [inline retrofit/src/main/java/retrofit/FormEncodingBuilder.java]: Should we add `isEmpty()` ?

File: retrofit/src/test/java/retrofit/RequestBuilderTest.java [hunk 2]
```diff
@@ -77,39 +78,39 @@ Response method(@Body RequestBody body) {
...
+    }
+    try {
+      buildRequest(Example.class);
+      fail();
+    } catch (IllegalArgumentException e) {
+      assertThat(e).hasMessage("Example.method: Only one encoding annotation is allowed.");
+    }
+  }
+
+  @Test public void onlyOneEncodingIsAllowedFormEncodingFirst() {
```
Comment (swankjesse) [inline retrofit/src/test/java/retrofit/RequestBuilderTest.java]: Is annotation order retained?

Comment (JakeWharton) [inline retrofit/src/test/java/retrofit/RequestBuilderTest.java]: Yes. We previously only asserted (and tested) one of these cases in the code before it was discovered that it matters.

File: retrofit/src/test/java/retrofit/RequestBuilderTest.java [hunk 3]
```diff
@@ -1404,186 +1405,186 @@ Response method(@Part("ping") RequestBody ping) {
...
+  @Test public void formEncodedFieldList() {
+    class Example {
+      @FormUrlEncoded //
+      @POST("/foo") //
+      Response method(@Field("foo") List<Object> fields, @Field("kit") String kit) {
+        return null;
+      }
+    }
+
+    List<Object> values = Arrays.<Object>asList("foo", "bar", null, 3);
```
Comment (swankjesse) [inline retrofit/src/test/java/retrofit/RequestBuilderTest.java]: Nice.

**Comment by swankjesse** [comment]
LGTM

File: retrofit/src/main/java/retrofit/FormEncodingBuilder.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (C) 2014 Square, Inc.
```
Comment (mattprecious) [inline retrofit/src/main/java/retrofit/FormEncodingBuilder.java]: 2015?

Comment (JakeWharton) [inline retrofit/src/main/java/retrofit/FormEncodingBuilder.java]: This is copy/pasted from OkHttp.

File: retrofit/src/main/java/retrofit/FormEncodingBuilder.java (hunk 1 above)
Comment (JakeWharton) [inline retrofit/src/main/java/retrofit/FormEncodingBuilder.java]: https://github.com/square/okio/pull/140

File: retrofit/src/main/java/retrofit/RequestBuilder.java [hunk 5]
```diff
@@ -74,13 +77,13 @@
     }
 
     switch (methodInfo.requestType) {
-      //case FORM_URL_ENCODED:
+      case FORM_URL_ENCODED:
       //  // Will be set to 'body' in 'build'.
```
Comment (jacobtabak) [inline retrofit/src/main/java/retrofit/RequestBuilder.java]: Extra `//` here

File: retrofit/src/main/java/retrofit/RequestBuilder.java [hunk 6]
```diff
@@ -381,6 +379,12 @@ Request build() {
     }
 
     RequestBody body = this.body;
+    if (formEncodingBuilder != null) {
```
Comment (jacobtabak) [inline retrofit/src/main/java/retrofit/RequestBuilder.java]: Smells a bit weird that this is an `if`/`else` block, as if form encoding has higher priority than multipart, when in reality they're mutually exclusive.

Comment (JakeWharton) [inline retrofit/src/main/java/retrofit/RequestBuilder.java]: Their mutual exclusivity is why it's an `else`. If one matches I don't want to pay the cost of the field lookup for the other.

Comment (JakeWharton) [inline retrofit/src/main/java/retrofit/RequestBuilder.java]: What I really want is
```java
RequestBody body = this.body;
if (body == null) {
  FormEncodingBuilder formEncodingBuilder = this.formEncodingBuilder;
  if (formEncodingBuilder != null) {
    body = formEncodingBuilder.build();
  } else {
    MultipartBuilder multipartBuilder = this.multipartBuilder;
    if (multipartBuilder != null) {
      body = multipartBuilder.build();
    }
  }
}
```
but that's far too ugly.

**Pull request merged**